### PR TITLE
Refactor to use better cli command framework

### PR DESCRIPTION
**Pull request by captncraig**
This breaks up the command switching logic in main into explicit commands that can be run independently.

Uses https://github.com/urfave/cli as its cli library.

### CODE REVIEW

**Comment by tlimoncelli** [comment]
New requirement: This change can't break any TeamCity builds.  The existing commands must work, or TC must be reconfigured to use commands that work in the old and new flags.

Currently the commands used by TeamCity are:

```
go test -v -verbose -provider BIND
bin/dnscontrol-Linux -providers bind -bindtree "$MDIR" push
bin\dnscontrol.exe push
```

File: cmd/debugPreprocess.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	"fmt"
+
+	"github.com/StackExchange/dnscontrol/pkg/normalize"
+	"github.com/urfave/cli"
+)
+
+var debugPreprocessCommand = &cli.Command{
+	Name:  "debug-preprocess",
```
Comment (tlimoncelli) [inline cmd/debugPreprocess.go]: Suggestion: s/debug-preprocess/output-ir/g

File: cmd/debugJs.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+func exit(err error) error {
+	if err == nil {
+		return nil
+	}
+	return cli.NewExitError(err, 1)
+}
+
+var debugJSCommand = &cli.Command{
+	Name:  "debug-js",
+	Usage: "Execute javascript and print generated json",
```
Comment (tlimoncelli) [inline cmd/debugJs.go]: s/Execute javascript and print generated json/Output intermediate representation (IR) after JavaScript is executed but before any validation/normalization/g

File: cmd/debugJs.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+func exit(err error) error {
+	if err == nil {
+		return nil
+	}
+	return cli.NewExitError(err, 1)
+}
+
+var debugJSCommand = &cli.Command{
+	Name:  "debug-js",
```
Comment (tlimoncelli) [inline cmd/debugJs.go]: s/debug-js/debug-otto/g

File: cmd/debugPreprocess.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	Usage: "Run validation and normalization logic, and print resulting json",
```
Comment (tlimoncelli) [inline cmd/debugPreprocess.go]: Output intermediate representation (IR) after running validation and normalization logic.

File: build/validate/validate.go [hunk 5]
```diff
@@ -93,14 +96,70 @@ func getModifiedFiles() ([]string, error) {
 	return strings.Split(string(out), "\n"), nil
 }
 
+func checkDirectOutput() error {
```
Comment (tlimoncelli) [inline build/validate/validate.go]: Missing comment.

File: cmd/commands.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+	catUtils = "utility"
+)
+
+// Run will execute the CLI
+func Run(version string) error {
+	app := cli.NewApp()
+	app.Version = version
+	app.Name = "dnscontrol"
+	app.HideVersion = true
+	app.Usage = "dnscontrol is a compiler and dsl for managing cloud dns zones"
```
Comment (tlimoncelli) [inline cmd/commands.go]: "dnscontrol is a compiler and DSL for managing dns zones" (cap DSL and remove "cloud")

File: cmd/commands.go [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+	ExecuteDSLArgs
+	JSONFile string
+}
+
+func (args *GetDNSConfigArgs) flags() []cli.Flag {
+	return append(args.ExecuteDSLArgs.flags(),
+		cli.StringFlag{
+			Destination: &args.JSONFile,
+			Name:        "json",
+			Usage:       "file containing intermediate json",
```
Comment (tlimoncelli) [inline cmd/commands.go]: s/file containing intermediate json/Read IR directly from this file. Do not process dnsconfig.js/g

File: cmd/commands.go [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+type GetDNSConfigArgs struct {
+	ExecuteDSLArgs
+	JSONFile string
+}
+
+func (args *GetDNSConfigArgs) flags() []cli.Flag {
+	return append(args.ExecuteDSLArgs.flags(),
+		cli.StringFlag{
+			Destination: &args.JSONFile,
+			Name:        "json",
```
Comment (tlimoncelli) [inline cmd/commands.go]: Suggestion: This flag should be "-input" or "-ir".  A flag named "-json" usually means to change the default format to JSON, which might be confusing.

File: cmd/commands.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+	DevMode  bool
+}
+
+func (args *ExecuteDSLArgs) flags() []cli.Flag {
+	return []cli.Flag{
+		cli.StringFlag{
+			Name:        "js",
+			Value:       "dnsconfig.js",
+			Destination: &args.JSFile,
+			Usage:       "Javascript file containing dns config",
```
Comment (tlimoncelli) [inline cmd/commands.go]: Main config file: Describes DNS zones, records, providers and other info.

Can this flag be "-js" for backwards compatibility but also something more descriptive like "-config"?

File: cmd/commands.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+	Pretty bool
+	Output string
+}
+
+func (args *PrintJSONArgs) flags() []cli.Flag {
+	return []cli.Flag{
+		cli.BoolFlag{
+			Name:        "pretty",
+			Destination: &args.Pretty,
+			Usage:       "Pretty print json",
```
Comment (tlimoncelli) [inline cmd/commands.go]: Prettyprint JSON

File: cmd/commands.go [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+		},
+		cli.StringFlag{
+			Name:        "out",
+			Destination: &args.Output,
+			Usage:       "File to write json to",
```
Comment (tlimoncelli) [inline cmd/commands.go]: Set output file (default stdout)

File: cmd/commands.go [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+		cli.StringFlag{
+			Name:        "providers",
+			Destination: &args.Providers,
+			Usage:       `Providers to enable (comma seperated list); default is all. Can exclude individual providers from default by adding '"_exclude_from_defaults": "true"' to the credentials file for a provider`,
+			Value:       "",
+		},
+		cli.StringFlag{
+			Name:        "domains",
+			Destination: &args.Domains,
+			Usage:       `Comma seperated list of domain names to include`,
```
Comment (tlimoncelli) [inline cmd/commands.go]: spell check: separated

File: cmd/commands.go [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+			Value:       "",
+		},
+	}
+}
+
+func (args *FilterArgs) shouldRunProvider(p string, dc *models.DomainConfig, nonDefaultProviders []string) bool {
```
Comment (tlimoncelli) [inline cmd/commands.go]: Missing comment

File: cmd/commands.go [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+	}
+	for _, prov := range strings.Split(args.Providers, ",") {
+		if prov == p {
+			return true
+		}
+	}
+	return false
+}
+
+func (args *FilterArgs) shouldRunDomain(d string) bool {
```
Comment (tlimoncelli) [inline cmd/commands.go]: Missing comment

File: cmd/previewPush.go [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+		fmt.Fprintf(os.Stderr, "##teamcity[buildStatus status='SUCCESS' text='%d corrections']", totalCorrections)
+	}
+	out.Debugf("Done. %d corrections.\n", totalCorrections)
+	if anyErrors {
+		return fmt.Errorf("Completed with errors")
+	}
+	return nil
+}
+
+// InitializeProviders takes a creds file path and a DNSConfig object. Creates all providers with the proper types, and returns them.
```
Comment (tlimoncelli) [inline cmd/previewPush.go]: Comments should be limited to 80 cols.

File: cmd/previewPush.go [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+	Category: catMain,
+	Flags:    globalPreviewArgs.flags(),
+}
+
+// PreviewArgs contains all data/flags needed to run preview, independently of CLI
+type PreviewArgs struct {
+	GetDNSConfigArgs
+	GetCredentialsArgs
+	FilterArgs
+	Delay int
```
Comment (tlimoncelli) [inline cmd/previewPush.go]: PowerShell flags are missing:  -fakeps, -psout, -pslog

**Comment by tlimoncelli** [comment]
We need to add a "dnscontrol check" which is like preview, but never talks to the providers. IT needs a way to check their work without having creds.  A simple syntax check (plus normalize + validate) is sufficient.  With the old system running "dnscontrol check" did this and we need to retain that.

File: cmd/debugJs.go (hunk 3 above)
Comment (captncraig) [reply cmd/debugJs.go]: I'd rather not mention otto in the command. There are other js interpreters I may eveluate and try to plug in in the future. otto is just one implementation.

**Comment by captncraig** [comment]
Pretty sure that check case is covered by `output-ir`

**Pull request merged**